commit d7296e4e84666d50e23d35b967ada44b5724cd80 (refs/changes/24/4924/7)
Author: John Levon <john.levon@joyent.com>
Date:   2018-10-17T22:40:26+00:00 (12 months ago)
    
    OS-7283 ./configure should support multiple compilers
    OS-7284 bootstrap node.js should use a runpath for libstdc++
    OS-7286 illumos-extra should compile with newer GCCs
    Reviewed by: Robert Mustacchi <rm@joyent.com>
    Approved by: Jason King <jason.king@joyent.com>

diff --git a/Makefile b/Makefile
index bf1800c..cd18df8 100644
--- a/Makefile
+++ b/Makefile
@@ -17,10 +17,13 @@
 # To build everything just run 'gmake' in this directory.
 #
 
-BASE =		$(PWD)
+include $(CURDIR)/../../build.env
+
+BASE =		$(CURDIR)
 DESTDIR =	$(BASE)/proto
 
 ifeq ($(STRAP),strap)
+
 STRAPPROTO =	$(DESTDIR)
 SUBDIRS = \
 	cpp \
@@ -35,12 +38,14 @@ SUBDIRS = \
 	openssl1x \
 	perl
 
-ifeq ($(BUILD_EXTRA_GCC),yes)
-SUBDIRS +=	gcc6 gcc7
-STRAPFIX +=	gcc6 gcc7
+COMMA=,
+EXTRA_COMPILERS = $(subst $(COMMA), , $(SHADOW_COMPILERS))
+SUBDIRS +=	$(EXTRA_COMPILERS)
+STRAPFIX +=	$(EXTRA_COMPILERS)
 STRAPFIX_SUBDIRS=$(STRAPFIX:%=%.strapfix)
-endif
+
 else
+
 STRAPPROTO =	$(DESTDIR:proto=proto.strap)
 SUBDIRS = \
 	bash \
@@ -83,6 +88,7 @@ SUBDIRS = \
 	vim \
 	wget \
 	xz
+
 endif
 
 PATH =		$(STRAPPROTO)/usr/bin:/usr/bin:/usr/sbin:/sbin:/opt/local/bin
@@ -101,6 +107,10 @@ GITDESCRIBE = \
 
 TARBALL =	$(NAME)-$(BRANCH)-$(TIMESTAMP)-$(GITDESCRIBE).tgz
 
+LIBSTDCXXVER_4 = 6.0.13
+LIBSTDCXXVER_6 = 6.0.22
+LIBSTDCXXVER_7 = 6.0.24
+
 #
 # Some software (e.g., OpenSSL 0.9.8) is very particular about the Perl
 # interpreter used during the build.  This is the full path to the version
@@ -150,13 +160,14 @@ $(DESTDIR)/usr/gnu/bin/gas: FRC
 # gas.
 #
 ifeq ($(STRAP),strap)
-$(DESTDIR)/usr/gcc/4/bin/gcc: $(DESTDIR)/usr/gnu/bin/gas
-	(cd gcc4 && \
+
+$(DESTDIR)/usr/gcc/$(PRIMARY_COMPILER_VER)/bin/gcc: $(DESTDIR)/usr/gnu/bin/gas
+	(cd $(PRIMARY_COMPILER) && \
 	    PKG_CONFIG_LIBDIR="" \
 	    STRAP=$(STRAP) \
 	    $(MAKE) DESTDIR=$(DESTDIR) install strapfix)
 
-$(SUBDIRS): $(DESTDIR)/usr/gcc/4/bin/gcc
+$(SUBDIRS): $(DESTDIR)/usr/gcc/$(PRIMARY_COMPILER_VER)/bin/gcc
 	(cd $@ && \
 	    PKG_CONFIG_LIBDIR="" \
 	    STRAP=$(STRAP) \
@@ -173,8 +184,9 @@ $(STRAPFIX_SUBDIRS): $(SUBDIRS)
 	    $(MAKE) DESTDIR=$(DESTDIR) strapfix)
 
 else
+
 $(DESTDIR)/usr/bin/gcc: $(DESTDIR)/usr/gnu/bin/gas
-	(cd gcc4 && \
+	(cd $(PRIMARY_COMPILER) && \
 	    PKG_CONFIG_LIBDIR="" \
 	    STRAP=$(STRAP) \
 	    $(MAKE) DESTDIR=$(DESTDIR) install strapfix)
@@ -191,19 +203,20 @@ $(SUBDIRS): $(DESTDIR)/usr/bin/gcc
 endif
 
 
-install: $(SUBDIRS) gcc4 binutils
+install: $(SUBDIRS) $(PRIMARY_COMPILER) binutils
 
 fixup_strap: $(STRAPFIX_SUBDIRS)
 
-install_strap: $(SUBDIRS) gcc4 binutils fixup_strap
+install_strap: $(SUBDIRS) $(PRIMARY_COMPILER) binutils fixup_strap
 
 clean:
-	-for dir in $(SUBDIRS) gcc4 binutils; \
+	-for dir in $(SUBDIRS) $(PRIMARY_COMPILER) binutils; \
 	    do (cd $$dir; $(MAKE) DESTDIR=$(DESTDIR) clean); done
 	-rm -rf proto
 
 manifest:
-	cp manifest $(DESTDIR)/$(DESTNAME)
+	sed 's/$$LIBSTDCXXVER/$(LIBSTDCXXVER_$(PRIMARY_COMPILER_VER))/g' \
+	    manifest >$(DESTDIR)/$(DESTNAME)
 
 mancheck_conf:
 	cp mancheck.conf $(DESTDIR)/$(DESTNAME)
diff --git a/Makefile.defs b/Makefile.defs
index 231e99c..1d37aef 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -15,6 +15,9 @@
 
 BASE =		$(PWD)
 DESTDIR =	$(BASE)/../proto
+
+include $(BASE)/../../../build.env
+
 VER.32 =	$(VER)-32$(STRAP)
 VER.64 =	$(VER)-64$(STRAP)
 OBJ.32 =	$(VER.32)$(SEPARATE_BUILD:yes=.build)
@@ -26,8 +29,9 @@ FETCH_BASE =	https://us-east.manta.joyent.com/Joyent_Dev/public/illumos-extra
 
 ifeq ($(STRAP),strap)
 STRAPPROTO =	$(DESTDIR)
-GCC =		$(DESTDIR)/usr/gcc/4/bin/gcc
-GXX =		$(DESTDIR)/usr/gcc/4/bin/g++
+GCC =		$(DESTDIR)/usr/gcc/$(PRIMARY_COMPILER_VER)/bin/gcc
+GXX =		$(DESTDIR)/usr/gcc/$(PRIMARY_COMPILER_VER)/bin/g++
+LIBSTDCXX_RUNPATH = usr/gcc/$(PRIMARY_COMPILER_VER)/lib
 else
 STRAPPROTO =	$(DESTDIR:proto=proto.strap)
 GCC =		$(DESTDIR)/usr/bin/gcc
diff --git a/README.md b/README.md
index 6008fb0..7111b3e 100644
--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 This repository, illumos-extra, is a collection of software which falls
 into two categories: either it is an illumos build and/or run-time
 dependency or it is a piece of additional software that SmartOS uses.
-For example, the `gcc` and `binutils` directories are examples of
+For example, the `gcc*` and `binutils` directories are examples of
 illumos dependencies, while `node.js` and `lldp` are examples of extra
 pieces of software that SmartOS uses to form its core ram-disk.
 illumos-extra is a fundamental part of the SmartOS build process;
diff --git a/gnupg/Makefile b/gnupg/Makefile
index abbcf1b..b92fe59 100644
--- a/gnupg/Makefile
+++ b/gnupg/Makefile
@@ -18,10 +18,11 @@
 #
 # CDDL HEADER END
 #
-# Copyright (c) 2012, Joyent, Inc.  All rights reserved.
+# Copyright 2018 Joyent, Inc.
 #
 
 VER =	gnupg-1.4.11
+PATCHES = Patches/*
 
 include ../Makefile.defs
 
diff --git a/gnupg/Patches/0001-With-C99-we-need-to-explicitly-ask-for-the-old-style.patch b/gnupg/Patches/0001-With-C99-we-need-to-explicitly-ask-for-the-old-style.patch
new file mode 100644
index 0000000..80de9d9
--- /dev/null
+++ b/gnupg/Patches/0001-With-C99-we-need-to-explicitly-ask-for-the-old-style.patch
@@ -0,0 +1,30 @@
+From 0e72ad50e11376749d41fbb2cce7fe1d6478ba14 Mon Sep 17 00:00:00 2001
+From: John Levon <john.levon@joyent.com>
+Date: Thu, 6 Sep 2018 11:42:17 +0000
+Subject: [PATCH] With C99, we need to explicitly ask for the old-style inline
+ semantics
+
+---
+ mpi/mpi-inline.h | 4 ++++
+ 1 file changed, 4 insertions(+)
+
+diff --git a/mpi/mpi-inline.h b/mpi/mpi-inline.h
+index f5e84e8..bb72e4a 100644
+--- a/mpi/mpi-inline.h
++++ b/mpi/mpi-inline.h
+@@ -29,8 +29,12 @@
+ #define G10_MPI_INLINE_H
+ 
+ #ifndef G10_MPI_INLINE_DECL
++#ifdef __GNUC_STDC_INLINE__
++#define G10_MPI_INLINE_DECL  extern inline __attribute__ ((__gnu_inline__))
++#else
+ #define G10_MPI_INLINE_DECL  extern __inline__
+ #endif
++#endif
+ 
+ G10_MPI_INLINE_DECL  mpi_limb_t
+ mpihelp_add_1( mpi_ptr_t res_ptr, mpi_ptr_t s1_ptr,
+-- 
+2.14.1
+
diff --git a/manifest b/manifest
index 5e4caa3..03aa914 100644
--- a/manifest
+++ b/manifest
@@ -114,11 +114,11 @@ f usr/lib/iconv/amd64/UCS-2%UTF-8.so 0755 root bin
 f usr/lib/iconv/amd64/UCS-2LE%UTF-8.so 0755 root bin
 # gcc
 f usr/lib/amd64/libgcc_s.so.1 0555 root bin
-f usr/lib/amd64/libstdc++.so.6.0.13 0555 root bin
-s usr/lib/amd64/libstdc++.so.6=libstdc++.so.6.0.13
+f usr/lib/amd64/libstdc++.so.$LIBSTDCXXVER 0555 root bin
+s usr/lib/amd64/libstdc++.so.6=libstdc++.so.$LIBSTDCXXVER
 f usr/lib/libgcc_s.so.1 0555 root bin
-f usr/lib/libstdc++.so.6.0.13 0555 root bin
-s usr/lib/libstdc++.so.6=libstdc++.so.6.0.13
+f usr/lib/libstdc++.so.$LIBSTDCXXVER 0555 root bin
+s usr/lib/libstdc++.so.6=libstdc++.so.$LIBSTDCXXVER
 f usr/lib/libssp.so.0.0.0 0555 root bin
 s usr/lib/libssp.so.0=libssp.so.0.0.0
 f usr/lib/amd64/libssp.so.0.0.0 0555 root bin
diff --git a/mdb_v8/Makefile b/mdb_v8/Makefile
index 92eb96f..94ff7fe 100644
--- a/mdb_v8/Makefile
+++ b/mdb_v8/Makefile
@@ -38,6 +38,11 @@ CTF_LIBS = $(VER)/build/ia32/mdb_v8.so \
 #
 CFLAGS += -g
 
+#
+# To avoid "error: left shift of negative value"
+#
+CFLAGS += -Wno-shift-negative-value
+
 #
 # There are a couple of bugs in the currently released version of mdb_v8
 # that do not support parallelism, thus we explicitly do not include any
diff --git a/ncurses/Makefile b/ncurses/Makefile
index e7fd906..7ac15f9 100644
--- a/ncurses/Makefile
+++ b/ncurses/Makefile
@@ -31,6 +31,7 @@ AUTOCONF_PREFIX =	/usr/gnu
 AUTOCONF_OPTS += \
 	-C \
 	--with-shared \
+	--without-cxx-binding \
 	--without-normal
 
 AUTOCONF_OPTS.64 += \
diff --git a/node.js/Makefile b/node.js/Makefile
index e8edf57..0724322 100644
--- a/node.js/Makefile
+++ b/node.js/Makefile
@@ -18,7 +18,7 @@
 #
 # CDDL HEADER END
 #
-# Copyright (c) 2013, Joyent, Inc.  All rights reserved.
+# Copyright 2018 Joyent, Inc.
 #
 
 MAJOR_VER =	0.10
@@ -29,7 +29,19 @@ include ../Makefile.defs
 NODE_ROOT =	$(DESTDIR)/usr/node/$(MAJOR_VER)
 VERSIONJS =	$(NODE_ROOT)/node_modules/platform_node_version.js
 
-CFLAGS +=	-Wno-unknown-pragmas
+#
+# ../deps/v8/src/objects.h:5188:44: error: left operand of shift expression
+# '(-1 << 3)' is negative
+#
+CXXFLAGS +=	-fpermissive
+
+#
+# https://github.com/nodejs/node/issues/6272
+#
+# Aside from the issues listed there, this causes v8 to explode during init in
+# AdvanceSweeper.
+#
+CXXFLAGS += -fno-delete-null-pointer-checks
 
 #
 # Node's build system is super broken.  If we pass it LIBS at configure time,
@@ -46,8 +58,12 @@ CFLAGS +=	-Wno-unknown-pragmas
 #
 LDFLAGS +=	-lumem
 
+#
+# Don't ask how long it took to get the literal $ORIGIN into the compiler
+# invocation here...
+#
 ifeq ($(STRAP),strap)
-	LDFLAGS += $(SYSLIBDIRS:%=-R$(DESTDIR)/%)
+	LDFLAGS += $(SYSLIBDIRS:%=-R$(DESTDIR)/%) -R'$$\$$'ORIGIN/../../../../$(LIBSTDCXX_RUNPATH)
 endif
 
 AUTOCONF_OPTS += \
@@ -64,7 +80,7 @@ AUTOCONF_OPTS += \
 
 AUTOCONF_CFLAGS =	CFLAGS="$(CPPFLAGS) $(CFLAGS)"
 AUTOCONF_LIBS =
-AUTOCONF_ENV +=		CXXFLAGS="$(CPPFLAGS) $(CFLAGS)"
+AUTOCONF_ENV +=		CXXFLAGS="$(CPPFLAGS) $(CFLAGS) $(CXXFLAGS)"
 
 #
 # Jump through hoops to get the locally-run build tools to build with the
diff --git a/node.js/Patches/0001-Disable-the-signbit-hack-for-recent-gcc-which-does-m.patch b/node.js/Patches/0001-Disable-the-signbit-hack-for-recent-gcc-which-does-m.patch
new file mode 100644
index 0000000..586e1ec
--- /dev/null
+++ b/node.js/Patches/0001-Disable-the-signbit-hack-for-recent-gcc-which-does-m.patch
@@ -0,0 +1,40 @@
+From 87539f2888aa1c0c7e37e0a7bc8c6d08127e1c55 Mon Sep 17 00:00:00 2001
+From: John Levon <john.levon@joyent.com>
+Date: Wed, 5 Sep 2018 09:39:24 +0000
+Subject: [PATCH] Disable the signbit hack for recent gcc, which does make it
+ visible to v8
+
+---
+ deps/v8/src/platform-solaris.cc | 2 +-
+ deps/v8/src/platform.h          | 2 +-
+ 2 files changed, 2 insertions(+), 2 deletions(-)
+
+diff --git a/deps/v8/src/platform-solaris.cc b/deps/v8/src/platform-solaris.cc
+index 07718fe..2bd3c17 100644
+--- a/deps/v8/src/platform-solaris.cc
++++ b/deps/v8/src/platform-solaris.cc
+@@ -61,7 +61,7 @@
+ // It seems there is a bug in some Solaris distributions (experienced in
+ // SunOS 5.10 Generic_141445-09) which make it difficult or impossible to
+ // access signbit() despite the availability of other C99 math functions.
+-#ifndef signbit
++#if !defined(signbit) && __GNUC__ < 6
+ // Test sign - usually defined in math.h
+ int signbit(double x) {
+   // We need to take care of the special case of both positive and negative
+diff --git a/deps/v8/src/platform.h b/deps/v8/src/platform.h
+index 63a6e44..db1b2ec 100644
+--- a/deps/v8/src/platform.h
++++ b/deps/v8/src/platform.h
+@@ -44,7 +44,7 @@
+ #ifndef V8_PLATFORM_H_
+ #define V8_PLATFORM_H_
+ 
+-#ifdef __sun
++#if defined(__sun) && __GNUC__ < 6
+ # ifndef signbit
+ int signbit(double x);
+ # endif
+-- 
+2.14.1
+
diff --git a/node.js/Patches/0002-converting-from-std-nullptr_t-requires-static_cast.patch b/node.js/Patches/0002-converting-from-std-nullptr_t-requires-static_cast.patch
new file mode 100644
index 0000000..dc23366
--- /dev/null
+++ b/node.js/Patches/0002-converting-from-std-nullptr_t-requires-static_cast.patch
@@ -0,0 +1,25 @@
+From 22ae87c9780464c6e8d35fa8be77d3a5123afce2 Mon Sep 17 00:00:00 2001
+From: John Levon <john.levon@joyent.com>
+Date: Wed, 5 Sep 2018 18:25:21 +0000
+Subject: [PATCH] converting from std::nullptr_t requires static_cast<>
+
+---
+ deps/v8/src/stub-cache.cc | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/deps/v8/src/stub-cache.cc b/deps/v8/src/stub-cache.cc
+index 8490c7e..54947e1 100644
+--- a/deps/v8/src/stub-cache.cc
++++ b/deps/v8/src/stub-cache.cc
+@@ -1361,7 +1361,7 @@ Handle<Code> StubCompiler::GetCodeWithFlags(Code::Flags flags,
+                                             Handle<String> name) {
+   return (FLAG_print_code_stubs && !name.is_null())
+       ? GetCodeWithFlags(flags, *name->ToCString())
+-      : GetCodeWithFlags(flags, reinterpret_cast<char*>(NULL));
++      : GetCodeWithFlags(flags, static_cast<char*>(NULL));
+ }
+ 
+ 
+-- 
+2.14.1
+
diff --git a/nss-nspr/Patches/0001-turn-off-various-warnings-that-break-the-build.patch b/nss-nspr/Patches/0001-turn-off-various-warnings-that-break-the-build.patch
new file mode 100644
index 0000000..e3ce93c
--- /dev/null
+++ b/nss-nspr/Patches/0001-turn-off-various-warnings-that-break-the-build.patch
@@ -0,0 +1,23 @@
+From 2dbe76aed64d04ff840d90523f19b27433ac9c8f Mon Sep 17 00:00:00 2001
+From: John Levon <john.levon@joyent.com>
+Date: Wed, 5 Sep 2018 10:51:40 +0000
+Subject: [PATCH] turn off various warnings that break the build
+
+---
+ nss/coreconf/config.mk | 3 +++
+ 1 file changed, 3 insertions(+)
+
+diff --git a/nss/coreconf/config.mk b/nss/coreconf/config.mk
+index bb87ad3..0dd4591 100644
+--- a/nss/coreconf/config.mk
++++ b/nss/coreconf/config.mk
+@@ -221,3 +221,6 @@ $(error Setting NSS_ENABLE_TLS_1_3 and NSS_DISABLE_ECC isn't a good idea.)
+ endif
+ DEFINES += -DNSS_ENABLE_TLS_1_3
+ endif
++
++CFLAGS += -Wno-unused -Wno-int-in-bool-context
++CXXFLAGS += -Wno-unused -Wno-int-in-bool-context
+-- 
+2.14.1
+
diff --git a/openssl1x/engine_pkcs11/hw_pk11_pub.c b/openssl1x/engine_pkcs11/hw_pk11_pub.c
index 3914053..8c67770 100644
--- a/openssl1x/engine_pkcs11/hw_pk11_pub.c
+++ b/openssl1x/engine_pkcs11/hw_pk11_pub.c
@@ -1247,7 +1247,7 @@ EVP_PKEY *pk11_load_privkey(ENGINE* e, const char *privkey_id,
 	pkcs11_uri uri_struct;
 	CK_RV rv;
 	CK_BBOOL is_token = CK_TRUE;
-	CK_BBOOL rollback = CK_FALSE;
+	CK_BBOOL rollback __attribute__((__unused__)) = CK_FALSE;
 	CK_BYTE attr_data[8][MAXATTR];
 	CK_OBJECT_CLASS key_class = CKO_PRIVATE_KEY;
 	CK_OBJECT_HANDLE ks_key = CK_INVALID_HANDLE;	/* key in keystore */
diff --git a/screen/Makefile b/screen/Makefile
index 26d82ed..5ed4ace 100644
--- a/screen/Makefile
+++ b/screen/Makefile
@@ -18,7 +18,7 @@
 #
 # CDDL HEADER END
 #
-# Copyright 2016, Joyent, Inc.  All rights reserved.
+# Copyright 2018, Joyent, Inc.  All rights reserved.
 #
 
 VER =	screen-4.4.0
@@ -28,11 +28,11 @@ include ../Makefile.defs
 PATCHES =	Patches/*
 
 #
-# We set _XOPEN_SOURCE=500 so that screen can access additional fields in
+# We set _XOPEN_SOURCE=700 so that screen can access additional fields in
 # struct msghdr. pkgsrc also does this.
 #
 AUTOCONF_ENV += \
-	CFLAGS=-D_XOPEN_SOURCE=500
+	CFLAGS="-std=c99 -D_XOPEN_SOURCE=700"
 
 #
 # It is important to make sure that screen has a prefix of /usr. When screen
